From b3df85df3a204193c64bcae82899377fa6048b83 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Sat, 15 Sep 2018 02:18:37 +0800
Subject: [PATCH 4692/5242] MGS-4266 gpu: drm: dpu-blitter: refine mutex logic

commit  9bd8ff09d343dd16353379051f641aa64d4d0390 from
https://source.codeaurora.org/external/imx/linux-imx.git

retry logic cause high cpu load for multiple instances,
should use simple mutex to sync dpu blitter directly.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dpu/dpu-blit.c  |    9 ---------
 drivers/gpu/imx/dpu-blit/dpu-blit.c |   11 -----------
 drivers/gpu/imx/dpu-blit/dpu-blit.h |    1 -
 3 files changed, 21 deletions(-)

diff --git a/drivers/gpu/drm/imx/dpu/dpu-blit.c b/drivers/gpu/drm/imx/dpu/dpu-blit.c
index e8511bb..9d1aa7a 100644
--- a/drivers/gpu/drm/imx/dpu/dpu-blit.c
+++ b/drivers/gpu/drm/imx/dpu/dpu-blit.c
@@ -88,10 +88,7 @@ static int imx_drm_dpu_set_cmdlist_ioctl(struct drm_device *drm_dev, void *data,
 
 	dpu_be = bliteng->dpu_be;
 
-retry:
 	ret = dpu_be_get(dpu_be);
-	if (ret == -EBUSY)
-		goto retry;
 
 	cmd_nr = req->cmd_nr;
 	cmd = (u32 *)(unsigned long)req->cmd;
@@ -145,10 +142,7 @@ static int imx_drm_dpu_wait_ioctl(struct drm_device *drm_dev, void *data,
 
 	dpu_be = bliteng->dpu_be;
 
-retry:
 	ret = dpu_be_get(dpu_be);
-	if (ret == -EBUSY)
-		goto retry;
 
 	dpu_be_wait(dpu_be);
 
@@ -279,10 +273,7 @@ static int dpu_bliteng_suspend(struct device *dev)
 	if (dpu_bliteng == NULL)
 		return 0;
 
-retry:
 	ret = dpu_be_get(dpu_bliteng);
-	if (ret == -EBUSY)
-		goto retry;
 
 	dpu_be_wait(dpu_bliteng);
 
diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.c b/drivers/gpu/imx/dpu-blit/dpu-blit.c
index 65575db..417080d 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.c
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.c
@@ -199,13 +199,6 @@ void dpu_bliteng_set_dev(struct dpu_bliteng *dpu_be, struct device *dev)
 int dpu_be_get(struct dpu_bliteng *dpu_be)
 {
 	mutex_lock(&dpu_be->mutex);
-	if (dpu_be->inuse) {
-		mutex_unlock(&dpu_be->mutex);
-		return -EBUSY;
-	}
-
-	dpu_be->inuse = true;
-	mutex_unlock(&dpu_be->mutex);
 
 	return 0;
 }
@@ -213,10 +206,6 @@ int dpu_be_get(struct dpu_bliteng *dpu_be)
 
 void dpu_be_put(struct dpu_bliteng *dpu_be)
 {
-	mutex_lock(&dpu_be->mutex);
-
-	dpu_be->inuse = false;
-
 	mutex_unlock(&dpu_be->mutex);
 }
 EXPORT_SYMBOL(dpu_be_put);
diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.h b/drivers/gpu/imx/dpu-blit/dpu-blit.h
index 75215b2..a3fe172 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.h
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.h
@@ -25,7 +25,6 @@ struct dpu_bliteng {
 	void __iomem *base;
 	s32 id;
 	struct mutex mutex;
-	bool inuse;
 	s32 irq_store9_shdload;
 	s32 irq_store9_framecomplete;
 	s32 irq_store9_seqcomplete;
-- 
1.7.9.5

